### ISSUE TITLE
Home Assistant Not Showing Sensors in Fahrenheit

### ISSUE THREAD

**Issue by nielsonm236** [nielsonm236/NetMod-ServerApp#109]
@yozik04 I would like your review of this issue.

After Network Module 20220921 0050 I decided to update my year-old installs of the Home Assistant Core and OS. I didn't make note of exactly what those old release revisions were, but they updated to this:
Home Assistant Core 2022.9.5
Home Assistant Operating System 6.6
Note that there is already yet another set of updates (just 14 hours ago), but the issue I'm going to describe appears to remain in the new updates and is intentional in HA.
ISSUE: Being in the US I was displaying my Temperature Sensors in Fahrenheit. That continues to work in the Browser, but sometime between 2021 and 2022 something changed in HA and I can no longer get the Network Temperature Sensors to display as Fahrenheit values in HA.

After some digging I found Issue #71462 in the homeassistant / core repository (https://github.com/home-assistant/core/issues/71462). It boils down to a comment made in Issue #72625 (https://github.com/home-assistant/core/issues/72625) that says "the problem is that Home Assistant is no longer automatically converting temperatures for non temperature sensors." I think (but not sure) that the Temperature Sensor Config statements now need the following to allow HA to properly do C to F conversions:
      device_class: 'temperature'
      state_class: 'measurement'
I think it would actually be put in the template looking like this;
      "device_class":"temperature",
      "state_class":"measurement",

This adds 57 bytes to the temperature sensor payload template. This is not something I'm keen to do but if necessary I will. I'm not sure if I can actually remove the single tick marks as shown.

@yozik04 Am I on the right track with this?
Mike

**Comment by yozik04** [comment]
This is what we have today:
```
{
    "uniq_id": "3a7c01fb9b4e_temp_06473b8b6d00",
    "name": "heat2floor temp 06473b8b6d00",
    "~": "NetworkModule/heat2floor",
    "avty_t": "~/availability",
    "stat_t": "~/temp/06473b8b6d00",
    "unit_of_meas": "°C",
    "dev": {
        "ids": [
            "NetworkModule_3a7c01fb9b4e"
        ],
        "mdl": "HW-584",
        "mf": "NetworkModule",
        "name": "heat2floor",
        "sw": "20220921 0050"
    }
}
```

According to the issues you have sent `device_class` and `state_class` need to be added to correctly convert temperature units.

This is what we need (I used shortened versions of parameter names):
```
{
    "uniq_id": "3a7c01fb9b4e_temp_06473b8b6d00",
    "name": "heat2floor temp 06473b8b6d00",
    "~": "NetworkModule/heat2floor",
    "avty_t": "~/availability",
    "stat_t": "~/temp/06473b8b6d00",
    "unit_of_meas": "°C",
    "dev_cla": "temperature",
    "stat_cla": "measurement",
    "dev": {
        "ids": [
            "NetworkModule_3a7c01fb9b4e"
        ],
        "mdl": "HW-584",
        "mf": "NetworkModule",
        "name": "heat2floor",
        "sw": "20220921 0050"
    }
}
```

Shortened parameter names were taken from here:
https://www.home-assistant.io/docs/mqtt/discovery/#discovery-payload

I am not sure if `state_class` may be omitted. Definitelly it is better to have it. https://developers.home-assistant.io/docs/core/entity/sensor/#available-state-classes

I am now thinking if we can drop `unit_of_measurement` if we add `device_class`. Most probably Celcius is default for temperature device class. I need to test it on my setup. Will try to do it today.

**Comment by nielsonm236** [comment]
Thanks. I will start a branch to do some testing as well.

**Comment by nielsonm236** [comment]
Just FYI I found this link where a user has the opposite problem (wants to see Celcius, but always seeing Fahrenheit). The reply seems to indicate that he must include the unit_of_measurement so that HA knows the data being provided is in Celcius, then HA will know whether to convert it based on the Unit System selected in "Settings / System / General".

**Comment by nielsonm236** [comment]
Ah ... but I see what you are saying. If unit_of_measurement is left out HA may assume the unit is Celcius for incoming data ... whereas the use in the link has incoming data in F. Forgot to include the link https://community.home-assistant.io/t/sensor-temperature-mqtt-topic-in-fahrenheit/435311

**Comment by nielsonm236** [comment]
Did some testing. This only works if "unit_of_meas" is included. But the tests also reveal I've got some kind of payload size calculation problem, or perhaps a buffer overrun. With the "dev_cla" and "stat_cla" statements and a maximum length device_name the device goes unavailable. When I reduced the number of characters in the devicename to reduce the payload size it started working again. So I will look for a calculation error ... or buffer overrun. We did a good job of compressing everything but sometimes this kind of problem shows up due to lack of code space for error checking. I'll find it.

**Comment by nielsonm236** [comment]
We're over-running the length of the uip_buf transmit buffer ... but only by 15 bytes. I will have to think about this a bit.

**Comment by nielsonm236** [comment]
I believe the right way to fix it is to implement the code needed to break an MQTT message into a series of packets, thus accommodating some future MQTT packet growth that we are not anticipating now.

HOWEVER, after a bunch of hours looking at code I can tell this is going to be a test intensive change. The current implementation assumed that all individual MQTT messages would be small enough to fit in a 440 byte buffer (uip_buf). In fact we had about 40 bytes to spare in the buffer based on the largest MQTT message (the Temperature Sensor Auto Discovery message). Now we've added 54 bytes to that message. The change to allow multiple packets to be sent for one message can emulate the Browser/HTTP design which already does that. However, I'm concerned that as we compressed the MQTT code some aspects of error checking and retry over multiple packets may have been removed to reduce code size. So I need to walk through this very carefully.

IN THE MEANTIME, I think I can implement a workaround. @yozik04 comments to me about how I handle Temperature Sensor data gathering made me look closer at that implementation. I think I can recover about 40 bytes of RAM since I no longer need to figure out which Temperature Sensor was added or deleted. That 40 bytes can be used to make the uip_buf bigger, getting us past the current issue with a relatively simple set of code changes, but of course leaving us with the exposure to future message length changes.

After the workaround I can work on the "right way".

Mike

**Comment by yozik04** [comment]
I have not tested config messages as I promised you on Saturday. Do you still need my testing with HA? I understood that you have tried yourself. I use ºC myself and would need to spin up a separate instance to do the testing with ºF.

**Comment by nielsonm236** [comment]
@yozik04 I don't think you need to do the testing. I'm able to reproduce this very easily.
Yes, the workaround might be adequate for a long time. But I was surprised that HA made such a large change and worry they might do it again. So I think I have time to figure this out after the workaround but don't want to have to do the research all over again a year or two from now.

**Comment by nielsonm236** [comment]
I can finally release this fix. For the time being I implemented a larger uip_buf, but that provoked a difficult to find bug in POST processing (an unitialized variable that has been there since the first releases). Fixed in release 20221009 2141.
